lib/impl/refactorRules.js

```diff
@@ -0,0 +1,384 @@
+  // Invalid field rules with fields within repeating section are also added to this list.
+  var rulesFlaggedForDeletion = {};
+
+  var conn = connections.mongooseConnection;
+  var formModel = models.get(conn, models.MODELNAMES.FORM);
+  var fieldModel = models.get(conn, models.MODELNAMES.FIELD);
+  var fieldRuleModel = models.get(conn, models.MODELNAMES.FIELD_RULE);
+  var pageRuleModel = models.get(conn, models.MODELNAMES.PAGE_RULE);
+
+  formModel.findById(form._id.toString()).populate("pages", "-__v")
+    .populate("pageRules", "-__v")
+    .populate("fieldRules", "-__v")
+    .populate("dataSources.formDataSources", "_id name")//Populating Data Sources
+    .populate("dataTargets", "_id name")//Populating Data Targets
+    .select("-__v").exec(function(err, formResult) {
+      if (err) {
+        return cb(err);
+      }
+
+      //Now have a form with all pages, pageRules and fieldRules populated
+      formModel.populate(formResult, {
+        "path": "pages.fields",
+        "model": fieldModel,
+        "select": "-__v"
+      }, function(err, updatedFormResult) {
+        if (err) {
+          return cb(err);
+        }
+
+        form = updatedFormResult;
+
+        formPages = updatedFormResult.pages;
+
+        async.parallel([
+          function(fieldRuleCb) {
+            async.each(formResult.fieldRules, async.apply(scanForInvalidFields, models.MODELNAMES.FIELD_RULE), fieldRuleCb);
+          },
+          function(pageRuleCb) {
+            async.each(formResult.pageRules, async.apply(scanForInvalidFields, models.MODELNAMES.PAGE_RULE), pageRuleCb);
+          }
+        ], function(err) {
+          if (err) {
+            return cb(err);
+          }
+
+          checkRulesWithRepeatingSections();
+
+          //Update the form object
+          async.series([
+            updateFormRules,
+            updateOrDeleteRules
+          ], cb);
+        });
+      });
+    });
+
+
+  function scanForInvalidFields(type, rule, callback) {
+    var targetField = rule.targetPage ? rule.targetPage : rule.targetField;
+    /**
+     * A field target is not valid if it does not exist in the form or it is an adminOnly field.
+     * @param fieldOrPageIdToCheck
+     * @param pages
+     * @returns {boolean}
+     */
+    function isFieldStillValidTarget(fieldOrPageIdToCheck, pages, cb) {
+      var fieldExistsInForm = false;
+      var invalidPages = _.filter(pages, function(page) {
+        var currentPageId = page._id.toString();
+
+        if (currentPageId === fieldOrPageIdToCheck) {
+          fieldExistsInForm = true;
+        }
+        var invalidFieldList = _.filter(page.fields, function(field) {
+          var currentFieldId = field._id.toString();
+          if (currentFieldId === fieldOrPageIdToCheck) {
+            //Current field exists
+            fieldExistsInForm = true;
+
+            //Field is admin only, therefore it is an invalid target for a rule.
+            if (field.adminOnly) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Replace this if-then-else statement by a single return statement. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://metrics.skunkhenry.com/coding_rules#rule_key=javascript%3AS1126)